File: airfoil_code/airfoil_dub_pan.jl
include("airfoil_util.jl")

function constDubPan(mu, p1, p2, p)
    #mu = doublet strength
    #p = point to calc velocity
    #p1 = start point of doublet panel
    #p2 = end point of doublet panel

    if p2 === nothing # wake
        theta = 0
        x1 = 0
        x = p[1] - p1[1]
        z = p[2] - p1[2]
        r_2 = (x-x1)^2+z^2
        u = mu/2/pi/r_2 * z
        w = -mu/2/pi / r_2 * x
    else
        theta = -atan(p2[2]-p1[2], p2[1]-p1[1]);
        p_new = coordRot2D(p, theta, p1)

        x1 = 0
        x2 = ptDist(p1, p2)

        x = p_new[1]
        z = p_new[2]
        if abs(z) < 0.00001 && x1<x && x<x2 # Influence on itself
            u = 0
            w = -mu/2/pi * (1/(x-x1) - 1/(x-x2))
        else
            u = mu/2/pi * (z/((x-x1)^2+z^2) - z/((x-x2)^2+z^2))
            w = -mu/2/pi * ((x-x1)/((x-x1)^2+z^2) - (x-x2)/((x-x2)^2+z^2))
        end
    end
    # transform back
    vel = [u, w]
    vel = coordRot2D(vel, -theta, [0,0])
    return vel

end

function airfoil_constDub(pan_pts, alpha)
pan_pts, c_pts, thetas, norms, tangents, dists = procPanels(pan_pts)
# append a panel for the wake
pan_pts = [pan_pts; [pan_pts[end,3], pan_pts[end,4], NaN, NaN]']

# conditions
U = 1
chord = 1
rho = 1.225

# Initialize solver matrix
num_pan = size(pan_pts,1) -1
A = zeros(num_pan+1, num_pan+1)
RHS = zeros(num_pan+1)
u_vec = U .* [cosd(alpha), sind(alpha)]

# Influence Matrix
for i in 1:num_pan
    RHS[i] =- u_vec'norms[i, :]
    for j in 1:num_pan+1
        if j == num_pan+1 # wake, last panel
            uw = constDubPan(1, pan_pts[j,1:2], nothing, c_pts[i,:])
        else
            uw = constDubPan(1, pan_pts[j,1:2], pan_pts[j,3:4], c_pts[i, :])
        end
        A[i,j] = uw'norms[i,:]
     end
end

# Kutta condition
A[end, 1] = 1
A[end, end] = 1
A[end, end-1] = -1
RHS[end] = 0

# remove one panel
remIdx = Int(round(size(pan_pts,1)/2))
num_pan = num_pan -1
keep_idx = [1:remIdx-1; remIdx+1:num_pan+2]

RHS = RHS[keep_idx]
A = A[keep_idx, keep_idx]
c_pts = c_pts[keep_idx[1:end-1], :]
norms = norms[keep_idx[1:end-1],:]
tangents = tangents[keep_idx[1:end-1],:]
thetas = thetas[keep_idx[1:end-1]]
pan_pts = pan_pts[keep_idx, :]
dists = dists[keep_idx[1:end-1]]

mu = A\RHS

pan_vels = zeros(num_pan, 2)
# velocities at each panel
for i in 1:num_pan
    pan_vels[i,:] = u_vec
    # perterbation velocity of panel on itself
    if i == 1  # First panel
        v_pan = 0.5*(mu[i+1]-mu[i])/ptDist(c_pts[i,:], c_pts[i+1,:])
    elseif i == num_pan  # last panel
        v_pan = 0.5*(mu[i]-mu[i-1])/ptDist(c_pts[i-1,:], c_pts[i,:])
    else
         v_pan = 0.5*(mu[i+1]-mu[i-1])/ptDist(c_pts[i-1,:], c_pts[i+1,:])
    end
    pan_vels[i, :] = pan_vels[i,:] +  coordRot2D([v_pan, 0], -thetas[i], [0,0])
    
    # influence of other panels
    for j in 1:num_pan+1
        if j == num_pan+1 # wake, last panel
            uw = constDubPan(mu[j], pan_pts[j,1:2], nothing, c_pts[i,:])
        else
            uw = constDubPan(mu[j], pan_pts[j,1:2], pan_pts[j,3:4], c_pts[i, :])
        end
        pan_vels[i, :] = pan_vels[i,:] + uw
     end
end

cp = 1 .- (pan_vels[:,1].^2 + pan_vels[:,2].^2) ./ U.^2
cpPlot = plot(c_pts[:,1], cp, yflip=true)

ps = cp.*0.5.*rho.*U.^2
fs = ps.*dists.*(-1).*norms[:,2]
cl = sum(fs)./0.5./chord./rho./U.^2
return [cl, cpPlot]
end


# Testing
pan_pts = readAF("airfoil.csv", true)
writeAF(pan_pts, "airfoil.dat")
#pan_pts = repanel(pan_pts, 50, 1, true)
aoa = 5
cl, cpPlot = airfoil_constDub(pan_pts, aoa)

println("CL: $cl")
display(cpPlot)

File: airfoil_code/airfoil_dub_pan.py
import numpy as np
import matplotlib.pyplot as plt
import math
from airfoil_util import *


def proc_panels(pts, debug=False):
    # Processes panels, gives collocation points, orientations, tangents, and normals
    # Points must be numpy array
    # Computes collocation points for panels defined by list of points
    co_pts = np.array([np.mean(pts[i:i+2,:], axis=0) for i in range(pts.shape[0]-1)])
    # computes panel orientation angles a for panels defined by list of points
    # positive going clockwise, starting at -x axis
    dz = np.diff(pts[:,1])
    dx = np.diff(pts[:,0])
    theta = np.arctan2(dz, dx)
    s_theta = -np.sin(theta)
    c_theta = np.cos(theta)
    norms = np.transpose(np.array([s_theta, c_theta]))
    tans = np.transpose(np.array([c_theta, -s_theta]))

    #tangent vector but aligned with free stream
    orients = [math.atan2( (pts[i,1]-pts[i+1,1])  ,  (pts[i+1, 0]-pts[i, 0]))   for i in range(pts.shape[0]-1)]
    tan_fs = [np.array([math.cos(a), -math.sin(a)]) for a in orients]
    tan_fs = np.array(tan_fs)
    lens = np.sqrt(dx**2 + dz**2)
    if debug:
        # DEBUG: show normals and tangents
        plt.plot(pts[:,0], pts[:,1], 'k', marker='o')
        plt.plot(co_pts[:,0], co_pts[:,1], marker='x')
        for i in range(len(pts)-1):
            m = 0.1
            plt.plot([co_pts[i][0],co_pts[i][0]+norms[i][0]*m], [co_pts[i][1],co_pts[i][1]+norms[i][1]*m], 'r')
            plt.plot([co_pts[i][0],co_pts[i][0]+tans[i][0]*m], [co_pts[i][1],co_pts[i][1]+tans[i][1]*m], 'b')
        plt.gca().set_aspect('equal', adjustable='box')
        plt.show()

    return co_pts, norms, tans, tan_fs, lens

def dub2D(mu, p1, p2, p):
    # p1 is set as the coordinate for doublet origin
    # p2 then rotated to land on the doublet x axis
    if p2 is None:
        # assume that theta is 0 as this is a wake
        theta = 0
        # x2 -> Inf, z -> 0
        x1 = 0
        x = p[0] - p1[0]
        z = p[1] - p1[1]
        r_2 = (x-x1)**2+z**2
        u = mu/2/math.pi/r_2 * z
        w = -mu/2/math.pi / r_2 * (x)
        uw = np.array([u,w])
    else:
        # coordinate transform
        theta = math.atan2(p2[1]-p1[1] , p2[0]-p1[0])
        theta = -theta
        t1 = np.array([[math.cos(theta), -math.sin(theta)], [math.sin(theta), math.cos(theta)]])
        # take p1 as coordinate of the origin of the dublet
        t2 = p - p1
        p_new = np.matmul(t1,t2)
        
        x1 = 0
        x2 = pt_dist(p1, p2)
        x = p_new[0]
        z = p_new[1]
        if abs(z) < 0.000001 and (x1 < x and x < x2):  # Influence on itself 
            u = 0
            w = -mu/2/math.pi * (1/(x-x1) - 1/(x-x2))
        elif abs(z) < 0.000001 and (x1 == x or x == x2):
            raise Exception('Singularity')
        else:
            u = mu/2/math.pi * (z/((x-x1)**2+z**2) - z/((x-x2)**2+z**2))
            w = -mu/2/math.pi * ((x-x1)/((x-x1)**2+z**2) - (x-x2)/((x-x2)**2+z**2))
            
        uw = np.array([u,w])
        # coordinate transform back into global frame
        t1 = np.array([[math.cos(theta), math.sin(theta)], [-math.sin(theta), math.cos(theta)]])
        uw = np.matmul(t1,uw)

    return uw


U = 1
chord = 1
alfa = 5
alfa = math.radians(alfa)
u_vec = U * np.array([math.cos(alfa), math.sin(alfa)])
roh = 1.225
#pts = read_dat('airfoil.dat')
pts = read_csv('airfoil.csv')
#pts = repanel(pts, 50, chord_dist = 'cosine', cos_wgt=1.0, show=True)
co_pts, norms, tans,  tan_fs, lens = proc_panels(pts, debug=False)

# Initialize matrices
num_pan = len(pts) - 1
a = np.zeros([num_pan+1, num_pan+1])
b = np.copy(a)
rhs = np.zeros(num_pan+1)

# Set kutta condition
a[-1,0] = 1
a[-1, -2] = -1
a[-1, -1] = 1
rhs[-1] = 0
for i in range(num_pan):
    rhs[i] = - np.matmul(u_vec, norms[i])
    for j in range(num_pan + 1):
        if j == num_pan: # wake
            uw = dub2D(1, pts[num_pan], None, co_pts[i])
        else:
           uw = dub2D(1, pts[j], pts[j+1], co_pts[i])
        a[i,j] = np.matmul ( np.transpose(uw), norms[i])
        b[i,j] = np.matmul(np.transpose(uw), tans[i])
# Remove one panel

rem_idx = int(round(num_pan/2)) -1
rhs = np.delete(rhs, rem_idx)
a = np.delete(a, rem_idx, 0)
a = np.delete(a, rem_idx, 1)
b = np.delete(b, rem_idx, 0)
b = np.delete(b, rem_idx, 1)
norms = np.delete(norms, rem_idx, 0)
tans = np.delete(tans, rem_idx, 0)
tan_fs = np.delete(tan_fs, rem_idx, 0)
pts = np.delete(pts, rem_idx, 0)
co_pts = np.delete(co_pts, rem_idx, 0)
lens = np.delete(lens, rem_idx, 0)

num_pan = num_pan - 1

# solve
mu = np.linalg.solve(a, rhs)
# Induced tangential velocities at each collocation point
q_ti = np.matmul(b, mu)
q_ti = q_ti[:-1]  # dump q_ti of wake
# At each local tangent velocity, add freestream contribution and panel contribution on itself
#11.38
for i in range(num_pan):
    
    if i == 0:  # First panel
        v_pan = 0.5*(mu[i+1]-mu[i])/pt_dist(co_pts[i], co_pts[i+1])
    elif i == num_pan - 1:  # last panel
        v_pan = 0.5*(mu[i]-mu[i-1])/pt_dist(co_pts[i-1], co_pts[i])
    else:
         v_pan = 0.5*(mu[i+1]-mu[i-1])/pt_dist(co_pts[i-1], co_pts[i+1])
    
    q_ti[i] = q_ti[i] + v_pan + np.dot(u_vec, tan_fs[i])
    
    #q_ti[i] = q_ti[i] + np.dot(u_vec, tan_fs[i])
# calculate Cps
cl = 0
cp = 1 - q_ti**2/U**2
ps = cp*0.5*roh*U**2
fs = 0
for i in range(len(ps)):
    f = ps[i]*lens[i]*-1*norms[i,1]
    fs = fs + f
cl = np.sum(fs)/(1/2*roh*chord*U**2)
print('Cl: {:.6f}'.format(cl))

plt.plot(co_pts[:,0], cp)
plt.grid(True)
plt.gca().invert_yaxis()
plt.ylabel('Cp')
plt.xlabel('x')
plt.show()